### PR TITLE
fix ReceiveMessageFromPacketInfo in DualMode ReceiveMessageFromAsync

### LINKED ISSUE #35296: Incorrect address being returned by ReceiveMessageFromAsync
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/content/problem/988661/incorrect-address-being-returned-by-receivemessage.html)._

---
.NET Core App - Version v=2.0

Summary:
I am doing a call to ReceiveMessageFromAsync in order to get the local address to which a UDP packet is sent.  If I use a DualMode socket and I send an IPv4 multicast message to the server, the server reports the ReceivedMessageFromPacketInfo.Address as being an IPv6 address, but the address bytes themselves are for an IPv Address.  Thus instead of getting 224.0.1.180, I am getting an address strin gof e000:01b4:800::.

Server Setup:
Open a socket on AddressFamily.InterNetworkV6
Set the DualMode flag on the socket to true
Bind the socket to (IPAddress.IPv6Any, port=5683)
Walk the list of all network interfaces and do a Join on appropriate interfaces for the address 224.0.1.180
Walk the list of all network interfaces and do a Join on appropriate interfaces for the address [FF02::DD]
On the socket call ReceiveMesageFromAsync
Receive the SocketAsyncEventArgs event
Example the ReceivedMessageFromPacketInfo.Address field

Client:
Open a socket on AddressFamily.InterNetwork
Bind the socket to (IPAddress.IPAny, port=0)
Send a UDP message to 22.0.1.180:5683

The correct address is returned for any of the following situations:
1. the client sends a UDP message to the IPV6 multicast address
2. The server does not use dual mode, but sets up independent sockets for IPv4 and IPv6 any addresses.

---
#### Visual Studio Feedback System on 4/14/2020, 02:20 AM: 

<p>We have directed your feedback to the appropriate engineering team for further evaluation. The team will review the feedback and notify you about the next steps.</p>

---
### Original Solutions
(no solutions)

### PR DESCRIPTION
**Pull request by wfurt**
It seems like this only impact async version. args.RemoteEndPoint seems correct but the PacketInfo got mangled. 
When socket is in DualMode we expect ControlDataIPv6 but we can get either ControlDataIPv6 or ControlData since we asked for both in SetReceivingPacketInformation()

I was expecting V4Mapped address and I tried to fix it to match Linux behavior. However that broke some other tests as we expect plain IPv4 address even on DualMode IPv6 socket. To keep this simple and to avid regression,  ReceiveMessageFromPacketInfo reports valid v4 or v6 address provided by OS. RemoteEndPoint respects socket type and always reports v4 mapped address. 

fixes #35296

### CODE REVIEW

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs:495]:
```suggestion
            else if (ipv4 && (_controlBufferPinned == null || _controlBufferPinned.Length != sizeof(Interop.Winsock.ControlData)))
```

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs
Comment (stephentoub) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs:498]: Why did you swap the order of these checks?

As long as we're changing it, we may as well eliminate the extra branch (there's no need to do the other comparison if ipv6 is true).

Comment (stephentoub) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs:498]: > Why did you swap the order of these checks?

Oh, I see, to prioritize to the address family check over the IsIPv4MappedToIPv6 check.

Comment (wfurt) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs:498]: yes, Since this is connection-less I want to make sure we have enough space for either v4 or v6.

**Comment by karelz** [comment]
@wfurt what is this blocked on? Do you want review from @antonfirsov?

**Comment by wfurt** [comment]
review would be good @karelz.

**Comment by karelz** [comment]
Good, @antonfirsov had it on his list ...

**Comment by karelz** [comment]
Looks like it requires deeper Sockets knowledge - @geoffkizer do you think you can help with the code review?

File: src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:440]: Why is this necessary? It looks like the callers already check this and call the right version of GetIPPacketInformation.

Comment (wfurt) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:440]: not really. We check for the socket type and we pre-allocate the buffer. But for dual mode we can get get either one when the receive operation is finished. Pure IPv4 or Ipv6 does not have this problem as the controls buffer will always match the protocol family.

Comment (wfurt) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:440]: BTW this check the length returned inside of the structure, not size of structure it self (like we do in some other places)

Comment (geoffkizer) [reply src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs:440]: Ah, that's what confused me. Callers are checking the size of the structure itself, not the returned length.

**Review by geoffkizer — APPROVED** [review]

**Pull request merged**
